CDAP-8342 Perform add/delete partition as program container user

rohitsinha54 opened the pull request:
- Sets the program container principal and keytab in explore query headers
- addParition/deleteParition operations are performed with the given impersonation information.
- If impersonation information for the above operation as it can be in case if kerberos disabled cluster Impersonator gets the current user ugi
- If no impersonation information given for the above call on kerberos enabled cluster then the operation will fail. 

Issue: https://issues.cask.co/browse/CDAP-8342
Build: http://builds.cask.co/browse/CDAP-RUT581-1

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:
```diff
@@ -166,7 +169,21 @@ public void initialize(TwillContext context) {
+        String keytabURI = programOpts.getArguments().getOption(ProgramOptionConstants.KEYTAB_URI);
+          Preconditions.checkArgument(!(Strings.isNullOrEmpty(principal) || Strings.isNullOrEmpty(keytabURI)));
+        impersonationInfo = new ImpersonationInfo(principal, keytabURI);
+        LOG.trace("Kerberos is enabled. Created impersonation info {} from program options.", impersonationInfo);
```
the keytabURI is a path available on all master nodes. But this will be in a program container, right? So is the keytabURI valid?

Commit:
CDAP-8342 Perform add/delete partition as program container user

rohitsinha54 commented:
The PR has been updated to pass the ProgramId from the program container. The old PR was thoroughly tested. Will need to test the changes again. This testing isn't done yet. I am working on it. 

- Removed passing the keytab URI in security headers
- Principal is still passed in security headers from program container to explore container
- Added a ProgramDiscoveryExploreClient which extends DiscoveryExploreClient and is aware of the ProgramId so that it can add it to the security header while making request.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:
```diff
@@ -83,7 +86,7 @@ public DistributedProgramRunnableModule(CConfiguration cConf, Configuration hCon
 
   // usable from any program runtime, such as mapreduce task, spark task, etc
-  public Module createModule() {
+  public Module createModule(@Nullable final ProgramId programId, @Nullable final String principal) {
```
I really don't like `null`. Why it will be `null` for program containers??

rohitsinha54 replied:
This is also called from DistributedMapReduceTaskContextProvider, SparkRuntimeContextProvider which does not need to pass this information correct ?

chtyim replied:
But why not?

chtyim commented on cdap-explore-client/src/main/java/co/cask/cdap/explore/client/ExploreHttpClient.java:
```diff
@@ -110,6 +110,18 @@ protected String getUserId() {
   }
 
+  @Nullable
+  protected String getUserPrincipal() {
```
This is not very good design. A better one is to have `protected void addExtraHeaders()` and by default is a no-op. It's much more flexible to have different headers added by different client.

Also, really try hard to not using `null`. It is really evil.

rohitsinha54 replied:
I will add it to https://github.com/caskdata/cdap/pull/7972/files#diff-818f9f36bd99e1e10bf573e17c876d68R120

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/auth/context/AuthenticationContextModules.java:
```diff
@@ -44,13 +53,26 @@ protected void configure() {
+   * if an principal was provided then that information is also included in the {@link Principal}.
    */
-  public AbstractModule getProgramContainerModule() {
+  public AbstractModule getProgramContainerModule(@Nullable final String principal) {
```
Why it is nulalble? When will it be `null`?

rohitsinha54 replied:
It can be null when kerberos is not enabled.

chtyim replied:
BTW, the signature should be returning `Module`, not `AbstractModule`.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/auth/context/AuthenticationContextModules.java:
```diff
@@ -44,13 +53,26 @@ protected void configure() {
     return new AbstractModule() {
       @Override
       protected void configure() {
-        bind(AuthenticationContext.class).to(ProgramContainerAuthenticationContext.class);
```
So the `ProgramContainerAuthenticationContext` become an unused class?? Why not move the logic below to `ProgramContainerAuthenticationContext`?

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/auth/context/AuthenticationContextModules.java:
```diff
@@ -44,13 +53,26 @@ protected void configure() {
+          public Principal getPrincipal() {
+            try {
+              return principal == null ?
+                new Principal(UserGroupInformation.getCurrentUser().getShortUserName(), Principal.PrincipalType.USER) :
```
Strictly speaking, shouldn't get the current user from here, as there is no guarantee who that will be. I think the UGI should be passed-in or determined when the `getProgramContainerModule` method is called:

```java
// BTW, that's how you avoid caller to pass in null
Module getProgramContainerModule() {
  String userName = UserGroupInformation.getCurrentUser().getShortUserName();
  return new ProgramContainerModule(new Principal(userName, Principal.PrincipalType.USER));
}

Module getProgramContainerModule(String principal) {
  String userName = UserGroupInformation.getCurrentUser().getShortUserName();
  return new ProgramContainerModule(new Principal(userName, Principal.PrincipalType.USER, principal));
}
```

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java:
```diff
@@ -117,6 +117,6 @@ protected void shutDown() throws Exception {
 
   private static Injector createInjector(CConfiguration cConf, Configuration hConf) {
-    return Guice.createInjector(new DistributedProgramRunnableModule(cConf, hConf).createModule());
+    return Guice.createInjector(new DistributedProgramRunnableModule(cConf, hConf).createModule(null, null));
```
That means, if a partition is created from a MapReduce task, then the program id and principal are not passed along? Seems wrong to me...

anew commented on cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java:
```diff
@@ -511,6 +518,19 @@ private static String isClassNotFoundException(Throwable e) {
     return null;
   }
 
+  @Nullable
```
Should these be done as @HeaderParam? What is the advantage of extracting them manually?

anew commented on cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java:
```diff
@@ -262,7 +262,7 @@ private static PluginInstantiator createPluginInstantiator(CConfiguration cConf,
 
   private static Injector createInjector(CConfiguration cConf, Configuration hConf) {
-    return Guice.createInjector(new DistributedProgramRunnableModule(cConf, hConf).createModule());
+    return Guice.createInjector(new DistributedProgramRunnableModule(cConf, hConf).createModule(null, null));
```
Similar to MapReduce, if a Spark program adds a partition, we must pass on the program id to make sure it is impersonated correctly.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillApplication.java:
```diff
@@ -37,7 +37,11 @@
+public abstract class
+
+
+
```
unintentional newlines?

anew reviewed (commented):
Minor comment and LGTM if tested on a cluster.

rohitsinha54 commented:
Tested on multinode cluster to work. Will merge after RUT.

rohitsinha54 commented:
Build passed here: http://builds.cask.co/browse/CDAP-RUT581-7
Will wait for ITM since the changes touches distribute program runnable and stuff: http://builds.cask.co/browse/CDAP-ITM11-2
